pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,424 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Tuple, Optional
+
+import torch
+
+from pytorch_lightning.metrics.utils import to_onehot, select_topk
+
+
+def _check_shape_and_type_consistency(preds: torch.Tensor, target: torch.Tensor) -> Tuple[str, int]:
+    """
+    This checks that the shape and type of inputs are consistent with
+    each other and fall into one of the allowed input types (see the
+    documentation of docstring of _input_format_classification). It does
+    not check for consistency of number of classes, other functions take
+    care of that.
+
+    It returns the name of the case in which the inputs fall, and the implied
+    number of classes (from the C dim for multi-class data, or extra dim(s) for
+    multi-label data).
+    """
+
+    preds_float = preds.is_floating_point()
+
+    if preds.ndim == target.ndim:
+        if preds.shape != target.shape:
+            raise ValueError(
+                "`preds` and `target` should have the same shape,",
+                f" got `preds` shape = {preds.shape} and `target` shape = {target.shape}.",
+            )
+        if preds_float and target.max() > 1:
+            raise ValueError(
+                "if `preds` and `target` are of shape (N, ...) and `preds` are floats, `target` should be binary."
+            )
+
+        # Get the case
+        if preds.ndim == 1 and preds_float:
+            case = "binary"
+        elif preds.ndim == 1 and not preds_float:
+            case = "multi-class"
+        elif preds.ndim > 1 and preds_float:
+            case = "multi-label"
```

Comment:
```suggestion
            case = "multi-dim"
```
or how the "multi-label" differs from "multi-class"

Reply:
This terminology is take from [sklearn](https://scikit-learn.org/stable/modules/multiclass.html), let me copy the relevant parts:
- Multiclass classification: classification task with more than two classes. Each sample can only be labelled as one class.
- Multilabel classification: classification task labelling each sample with x labels from n_classes possible classes, where x can be 0 to n_classes inclusive. This can be thought of as predicting properties of a sample that are not mutually exclusive. Formally, a binary output is assigned to each class, for every sample. 

The same thing in [wikipedia](https://en.wikipedia.org/wiki/Multi-label_classification) [articles](https://en.wikipedia.org/wiki/Multiclass_classification)

Reply:
, can we include it in the docs, this explanation, eventually with link to wiki :]

Reply:
ok, I see it now in the docs... sorry :]